packages/SwingSet/src/kernel/kernel.js

```diff
@@ -646,6 +681,12 @@ export default function buildKernel(kernelEndowments) {
     manager.setVatSyscallHandler(vatSyscallHandler);
   }
 
+  async function removeVatManager(vatID) {
+    const old = ephemeral.vats.get(vatID);
+    ephemeral.vats.set(vatID, harden({ dead: true }));
+    old.manager.shutdown();
```

Comment:
`shutdown()` returns a promise (killing a child process is async), I think you either want to `await` it, or just return it (and remove the unnecessary `async`)

Reply:
Returning the promise seems like the right thing, since `terminateVat` does a then anyway.

Reply:
Although (as I think we explored) we need to make sure that doesn't result in nondeterminism.. the userspace `.done` promise should be resolved independently of whenever `shutdown`'s promise resolves. If `shutdown` rejects for whatever reason, that might log a message or kill the kernel, but it shouldn't be revealed to the control facet.

Reply:
I'm pretty sure I didn't follow that.

Reply:
The `createVat()` caller gets back a record with both a `terminate` method and a `done` Promise, and the API says that `done` will be resolved when the vat is dead (whether via `terminate` or via metering). So we must arrange for `done` to be resolved, which happens by injecting a message to the `vatAdminVat` (which holds the resolver for `done`).

The potentially non-determistic approach would be to delay sending that message until the Promise returned by `shutdown` has resolved. The exact timing of `shutdown` will depend upon the kind of worker being used, as well as various behaviors of the host's (Linux) kernel scheduler, because it might kill a child process and then wait (briefly) for it to finish shutting down. We don't want host-by-host variations in this timing to cause some consensus nodes (validators) to deliver the `done` resolution at different times: that would be a consensus failure.

So we should queue the "please resolve `done`" message to vatAdminVat during `terminate`, but not tie it to the Promise we get back from `shutdown` (and therefore not tied to the promise we get back from `removeVatManager`). *That* promise should have a .catch on it which signals a problem of some sort, but it should not have a `.then` on it which is involved in any way with the `done` message.

Does that make more sense?

Reply:
Yes.  The `.then` in `terminateVat` will handle both forks (reject and resolve) of the shutdown, and so shouldn't leak any of that back to the user vat.  HOWEVER, entirely separate from any of the changes in this PR, looking through the code I don't see where that the `done` promise ever gets resolved.  By following the code I see that it's ultimately up to the `notifyTermination` function, but this as best I can determine doesn't seem to ever be called anywhere.

Reply:
As an addendum to that, adding the tests to verify that the proper temporal logic of terminate vs. done my be the thing that drives smushing the termination test stuff in with the vat admin test stuff that you wanted :-)

Reply:
yeah, I think when you merge my PR and then uncomment the `done()` part, you'll run into the part that needs to be addressed.

BTW just to be super clear, in addition to not leaking the did-it-reject or did-it-resolve of `shutdown`, it must also not leak exactly when the resolution happened, since for whatever reason it might take the (unix) kernel many seconds to kill off the child process. When you add the code that fires `done`, it shouldn't be causally linked to the outcome of `shutdown`.

And yeah, to implement `done` resolution, we need a way for `terminate` to reach that `notifyTermination` callback (added as a `managerOption` so it can be called when metering fails). Maybe `vatManager.shutdown` should call the callback first, and then do whatever it needs to do to kill off the worker. Or maybe we should add `notifyTermination` to the ephemeral vat record so it can be called from outside the manager. Not sure what the best answer will be.